### PR TITLE
Define differentiability variables for NN operators

### PR DESCRIPTION
**Pull request by rajeevnalawadi**

### CODE REVIEW

File: onnx/defs/nn/defs.cc [hunk 1]
```diff
@@ -525,19 +529,30 @@ ONNX_OPERATOR_SET_SCHEMA(
             "The indices are linear, i.e. computed considering the tensor as flattened 1-D tensor, "
             "assuming row-major storage. Also, the linear indices should not consider padding. "
             "So the values in indices are in the range [0, N x C x D1 x ... x Dn).",
-            "T2")
+            "T2",
+            OpSchema::Single,
+            true,
+            1,
+            OpSchema::Differentiable)
```
Comment (wschin) [inline onnx/defs/nn/defs.cc:536]: I feel it's not differentiable. Do you have a Pytorch example to double-check?

File: onnx/defs/nn/defs.cc [hunk 2]
```diff
@@ -525,19 +529,30 @@ ONNX_OPERATOR_SET_SCHEMA(
...
             2,
             "output_shape",
             "The shape of the output can be explicitly set which will cause pads values to be auto generated. If 'output_shape' is specified, "
             "'pads' values are ignored.",
             "T2",
-            OpSchema::Optional)
+            OpSchema::Optional,
+            true,
+            1,
+            OpSchema::Differentiable)
```
Comment (wschin) [inline onnx/defs/nn/defs.cc:546]: Shape shouldn't be differentiable.

File: onnx/defs/nn/defs.cc [hunk 3]
```diff
@@ -858,17 +881,29 @@ ONNX_OPERATOR_SET_SCHEMA(
             "in effect, the operation expects input data tensor "
             "to arrive with the dimension denotation of [DATA_BATCH, "
             "DATA_CHANNEL, DATA_FEATURE, DATA_FEATURE ...].",
-            "T1")
+            "T1",
+            OpSchema::Single,
+            true,
+            1,
+            OpSchema::Differentiable)
```
Comment (wschin) [inline onnx/defs/nn/defs.cc:888]: The tags for quantized ops make sense. Just out of curiosity, are they also differentiable in Pytorch?

File: onnx/defs/nn/defs.cc [hunk 4]
```diff
@@ -1024,7 +1086,11 @@ ONNX_OPERATOR_SET_SCHEMA(
...
```
Comment (wschin) [inline onnx/defs/nn/defs.cc:1093]: This one I am not sure. Conv itself is differentiable, but its type conflicts with my mental assumption --- integer tensors are not differentiable.

File: onnx/defs/nn/defs.cc [hunk 5]
```diff
@@ -1568,46 +1679,94 @@ ONNX_OPERATOR_SET_SCHEMA(
...
         .Output(
             1,
             "mean",
             "The running mean after the BatchNormalization operator.",
             "T",
-            OpSchema::Optional)
+            OpSchema::Optional,
+            true,
+            1,
+            OpSchema::Differentiable)
```
Comment (wschin) [inline onnx/defs/nn/defs.cc:1740]: It's differentiable in math. However, I checked Pytorch and TF batchnorm but none of them outputs "mean", "var", "saved_mean", "saved_var". May we make them non-differentiable?

File: onnx/defs/nn/defs.cc [hunk 6]
```diff
@@ -1731,17 +1934,39 @@ ONNX_OPERATOR_SET_SCHEMA(
             "If it's non-zero, output will be a random dropout of the scaled input, which is typically "
             "the case during training. It is an optional value, if not specified it will default to 0.5.",
             "T1",
-            OpSchema::Optional)
+            OpSchema::Optional,
+            true,
+            1,
+            OpSchema::Differentiable)
```
Comment (wschin) [inline onnx/defs/nn/defs.cc:1940]:
```suggestion
            OpSchema::NonDifferentiable)
```

File: onnx/defs/nn/defs.cc [hunk 7]
```diff
@@ -1731,17 +1934,39 @@ ONNX_OPERATOR_SET_SCHEMA(
...
+        .Output(
+            1,
+            "mask",
+            "The output mask.",
+            "T2",
+            OpSchema::Optional,
+            true,
+            1,
+            OpSchema::Differentiable)
```
Comment (wschin) [inline onnx/defs/nn/defs.cc:1969]:
```suggestion
            OpSchema::NonDifferentiable)
```

File: onnx/defs/nn/defs.cc [hunk 8]
```diff
@@ -1958,8 +2219,24 @@ ONNX_OPERATOR_SET_SCHEMA(
...
-        .Output(0, "Y", "Ngram results", "T1")
+        .Input(
+            0,
+            "X",
+            "Input for n-gram extraction",
+            "T",
+            OpSchema::Single,
+            true,
+            1,
+            OpSchema::Differentiable)
```
Comment (wschin) [inline onnx/defs/nn/defs.cc:2230]:
```suggestion
            OpSchema::NonDifferentiable)
```

File: onnx/defs/nn/defs.cc [hunk 9]
```diff
@@ -1958,8 +2219,24 @@ ONNX_OPERATOR_SET_SCHEMA(
...
+        .Output(
+            0,
+            "Y",
+            "Ngram results",
+            "T1",
+            OpSchema::Single,
+            true,
+            1,
+            OpSchema::Differentiable)
```
Comment (wschin) [inline onnx/defs/nn/defs.cc:2239]:
```suggestion
            OpSchema::NonDifferentiable)
```

File: onnx/defs/nn/defs.cc (hunk 6 above)
Comment (rajeevnalawadi) [reply onnx/defs/nn/defs.cc]: Updated with suggestion

File: onnx/defs/nn/defs.cc (hunk 7 above)
Comment (rajeevnalawadi) [reply onnx/defs/nn/defs.cc]: Updated as NonDifferentiable

File: onnx/defs/nn/defs.cc (hunk 8 above)
Comment (rajeevnalawadi) [reply onnx/defs/nn/defs.cc]: Updated as NonDifferentiable

File: onnx/defs/nn/defs.cc (hunk 9 above)
Comment (rajeevnalawadi) [reply onnx/defs/nn/defs.cc]: Updated as NonDifferentiable

File: onnx/defs/nn/defs.cc (hunk 5 above)
Comment (rajeevnalawadi) [reply onnx/defs/nn/defs.cc]: Made the Batchnorm variables NonDifferentiable

File: onnx/defs/nn/defs.cc (hunk 3 above)
Comment (rajeevnalawadi) [reply onnx/defs/nn/defs.cc:888]: Torch indicates these quantized variables as Differentiable

**Review by wschin — APPROVED** [review]
This PR is almost ready to be merged. The only unclear piece to me is ConvInteger's differentiable behavior. Do you have a Pytorch example I can play with? As mentioned in one of my comments, I feel ConvInteger is not differentiable, but I'd like to double-check with Pytorch code.

**Comment by rajeevnalawadi** [comment]
@Wei_sheng: 
Pytorch has 3 specific modes for quantization
For training case the “Quantization aware training” QAT is applicable as described in the below linked page
https://pytorch.org/docs/stable/quantization.html#quantization-workflows
Snippet from the page under Quantization workflows & QAT (#3) “…Computations will take place in FP32 but with values clamped and rounded to simulate the effects of INT8 quantization…”

Code for checking QAT
https://pytorch.org/tutorials/advanced/static_quantization_tutorial.html
